### ISSUE TITLE
[Config] Configurator UI

### ISSUE THREAD

**Issue by maidmaid** [symfony/symfony#13456]
**Goal : make each bundle configuration editable with a web page**, like the current ``app_dev.php/_configurator`` but better :smiling_imp:.

Here, an example for ``Symfony\Bundle\FrameworkBundle\DependencyInjection\Configuration`` ([full screen here](https://cloud.githubusercontent.com/assets/4578773/5792279/2ae556c6-9f11-11e4-8ab4-2bbca146b849.png)) :

![configurator-ui](https://cloud.githubusercontent.com/assets/4578773/5792276/0b88f9b8-9f11-11e4-9bf1-64ac8a070a28.png)

I think it would be a great benefit for Symfony. Indeed, currently developer have to check out the configuration settings in the documentation of the bundle (or running commands ``config:debug`` and ``config:dump-reference``), however ``ConfigurationInterface::getConfigTreeBuilder()`` contains all  required information. Why not create a GUI for this ``TreeBuilder`` ?

By using the ``Form`` component, it gets simpler. Here's the code I wrote in a controller to achieve the result, as shown above by the screenshot :

```php
/**
 * @Route("/_myconfigurator", name="configurator")
 */
public function configureAction()
{
	$configuration = new \Symfony\Bundle\FrameworkBundle\DependencyInjection\Configuration(true);
	$processor = new Processor();
	$config = $processor->processConfiguration($configuration, array());

	$builder = $this->createFormBuilderFromConfiguration($configuration, $config);

	return $this->render('default/index.html.twig', array('form' => $builder->getForm()->createView()));
}

/**
 * Returns a form builder from configuration
 * @param ConfigurationInterface $configuration
 * @param mixed $config
 * @return FormBuilderInterface
 */
protected function createFormBuilderFromConfiguration($configuration, $config)
{
	$tree = $configuration->getConfigTreeBuilder()->buildTree();
	$builder = $this->get('form.factory')->createNamedBuilder($tree->getName(), 'form', $config);

	return $this->createFormBuilderFromNode($tree, $builder);
}

/**
 * Returns a form builder from node
 * @param NodeInterface $tree
 * @param FormBuilderInterface $builder
 * @return FormBuilderInterface
 */
protected function createFormBuilderFromNode($tree, $builder)
{
	if (!$tree instanceof ArrayNode) {
		return;
	}

	foreach ($tree->getChildren() as $node) {
		$subBuilder = $this->createFormBuilderFromNode(
			$node,
			$this->get('form.factory')->createNamedBuilder($node->getName())
		);
		if ($subBuilder !== null) {
			$builder->add($subBuilder);
		} else {
			$builder->add($node->getName(), null, array('attr' => array('placeholder' => $node->getDefaultValue())));
		}
	}

	return $builder;
}
```

Of course, this is a draft. Here are some points that have been thinking of to improve and complete
the work :
  - save changes in the configuration file,
  - load configuration of each active bundle (``framework``, ``security``, ``twig``, ``AppBundle``...),
  - choose the right type of field for each type of configuation nodes :
    - ``ScalarNode`` represented by a text field,
    - ``IntegerNode`` a numeric field,
    - ...
  - do « smart » fields taking into account all the various constraints of the nodes and options : 
    - ``defaultValue()`` add a placeholder in the field,
    - ``isRequired()`` defines the field as mandatory,
    - ``cannotBeEmpty()`` add a check constraint ``NotBlank``,
    - ``getAttribute('info', '')`` add tooltip,
    - ``ifNotInArray()->thenInvalid()`` set up a combo,
    - ...
  - switch the environnement configuration (dev, prod...),
  - etc.

What do you think ?

**Comment by WouterJ** [comment]
-1 for Symfony

* We are developers, why would we bother using a GUI when we can edit perfectly readable Yaml files? (or perfectly autocompleted Xml files or perfectly familair PHP files?)
* This very simple set-up is indeed not hard. But when you want to do it 100% correct, you are going to hit many difficult points. Prototyped arrays, normalizations, custom validations, etc.

**Comment by jamesmoey** [comment]
This feel like belong to a third party bundle.

**Comment by kix** [comment]
IMO, this should be implemented in another way.
The problem is that configuration options of a bundle could be anything, as it's just YAML. So, in order to edit these values from a web interface we need to be able to configure what fields to use and how they should look.
So, what I'd suggest here is to make a separate bundle for these options that would compose a configuration form for all the bundles using a configurator service and a tag for dependencies that add options/values using a Form component-alike API.

**Comment by javiereguiluz** [comment]
@maidmaid I don't know if this feature is going to be included or not, but I love when developers propose new features for Symfony. Thanks for that!

Regarding your proposal, I actually see two problems:

**Comment by kix** [comment]
@javiereguiluz, well, what if we could declare the configuration like this, given `config_handler` is the service to be implemented?

```
$configHandler = $this->container->get('config_handler');
$configHandler
	->add('doctrine.dbal.host', 'text', array(
		'label' => 'Database host',
		'property_path' => 'database_host',
	))
	->add('doctrine.dbal.driver', 'choice', array(
		'label' => 'Database driver',
		'property_path' => 'database_driver',
		'choices' => array(
			'pdo_mysql' => 'MySQL',
			'pdo_pgsql' => 'PostgreSQL',
		),
	))
	// ...
;
```

Then, all we'd have to do is implement a controller that would render the form for these fields (I think this feature should reuse the Form component) and process POST requests, persisting the data into `parameters.yml`

**Comment by WouterJ** [comment]
How would that config handler know where to save it?

**Comment by kix** [comment]
@WouterJ, if we're talking plain arrays here (like in default symfony-standard distribution), I think `property_path`'s would suffice. Though if someone needs arrays in `parameters.yml`, that would be a bit harder. But I think that Symfony form component can handle that.

**Comment by WouterJ** [comment]
Parameters.yml is just a convention used by the symfony standard. Even app/config/config.yml is a convention of the SE.

**Comment by javiereguiluz** [comment]
@kix imagine an application that defines its application as follows:

```yaml
# app/config/config.yml
imports:
    - { resource: framework.yml }

# app/config/framework.yml
framework:
    default_locale:  "%locale%"
```

My questions:

  * How does this configurator know that the value of `default_locale` is not a real value but a parameter?
  * How does this configurator know where to save those parameters? (In this application it could be stored in a file different from `app/config/parameters.yml`)
  * How does this configurator know which options to store in the custom `app/config/framework.yml` file and which ones in other files?

To sum up, I've seen an incredible variety of application configurations in Symfony application, using lots of custom files, locations and formats. For that reason, in my opinion is "impossible" to create a generic configurator. Maybe there is a way to do this, but I can't see it.

**Comment by kix** [comment]
@javiereguiluz, thanks, that's a good sample. Well, I'll think about that a bit, and make a pull request/bundle if anything.

**Comment by kix** [comment]
@javiereguiluz, given your example:
1. If we configure the configurator during container compilation, we can dynamically specify the parameters we need to persist into `.yml` files. Take a look at my second comment in this thread.
2. We can pass filenames and paths as options to the configurator.
3. We don't need to distinguish editable and non-editable parameters, as we're just building a form that has fields for params we want to edit though the interface.
@maidmaid suggests to create the form builder from the configuration, and my proposal is to make it declarative. We don't want to dynamically generate a form for the whole configuration, we only want some of the fields.
Anyway, the code always speaks for itself, so I think that posting a link to a repo implementing these ideas would be the best answer here (I hope I'll have some spare time to do that)

**Comment by ogizanagi** [comment]
Why not doing the same thing, but only for .dist files, in order to set or edit parameters ?
Files locations could be retrieved from `composer.json` "incenteev-parameters" and following a given nomenclature in order to gather parameters by "namespace" could make the form more readable.

e.g: 

```yaml
#parameters.yml
    database_driver:   pdo_mysql
    database_host:     127.0.0.1
    [...]
    acme.demo_bundle.demo_service.enabled: true
    acme.demo_bundle.api_client.host: http://api.acme.demo.com
    acme.demo_bundle.api_client.api_key: 36546DRFUV46U64554YDFKUY
```

will result into the following form presentation:

````yaml
◎ Tab 'General'
    database_driver
    database_host

◎ Tab Acme
   - Section 'Demo Bundle'
       -Sub-section 'Demo Service'
            enabled
        - Sub-section 'Api client'
            host
            api_key

```

I'm aware that it is not exactly what is suggested here, but previous comments share some valid concerns about the configuration flexibility and disparity between projetcs and developers.
But I think there is a good idea to keep in this suggestion, and relying on `.dist` files and parameters set inside will allow developper to expose only necessary parameters, and without any concerns about the real configuration files.

But IMO, this has no interest in being part of the symfony core. Maybe in the `[sensiolabs/SensioDistributionBundle](https://github.com/sensiolabs/SensioDistributionBundle)` ? Or only a third-party bundle.

**Comment by zebba** [comment]
I'd vote -1 too.

From my point of view the configuration in app/config is meant to harness the behaviour of the different bundles. They are not meant to be parameters that you swap between by the minute just to see what something does.

There is no way to prevent serious problems from happening.
1. What happens if you put the content of security.yml into config.yml and mess around with the firewall? If you hit save you kicked yourself out of the game and have to fallback to manually editing the file again.
2. If you change the CSRF token from the GUI what happens to the key you currently use? From the current process the old token is valid upon submitting the data but there needs to be a container-refresh at some point before (!) the next output of a form.
3. If multiple developers change settings at the same time, how is this supposed to work?

There are a lot more concerns like this and the risks involved are not worth the benefits.

You will also have to rethink your file permissions at some point depending on your setup.

**Comment by Rvanlaak** [comment]
Nice plan from a DX perspective! Choosing for [PhpStorm with the Symfony plugin](https://plugins.jetbrains.com/plugin?pr=&pluginId=7219) will have a similar result as having a GUI. The IDE helps you really well in determining configuration options and default values.

> * switch the environnement configuration (dev, prod...),

For me this is something that happens in the Apache VirtualHost by setting an environment variable

**Comment by stof** [comment]
I see several issues for this (which are already true for the current configurator):

- using a form to configure bundles means that the bundle cannot have mandatory configuration, as you need to be able to boot Symfony to display the form
- this assumes that config files are writable by the webserver, which may not be true at all
- building a form to handle configs is a very hard task for a generic way (the Config component allows very complex setups). and if you force each bundle to provide their config form, you won't get a good support for the feature (and people having to write a form for complex configs like the DoctrineBundle or SecurityBundle ones will become crazy)

**Comment by fabpot** [comment]
Closing as this is not something we want to package at the core of Symfony.
